### DISCUSSION TITLE
Federation Question

### DISCUSSION THREAD

**Discussion by sgentile** [MichalLytek/typegraphql-prisma#192] [Q&A]
Since the generator is creating graphql schemas from Prisma, is there a best practice in terms of how to add references/keys/extends and then also adding dataloaders ?

Do you extend the schema and add the directives in extensions ?

**Comment by MichalLytek** [comment]
Such complex cases almost always requires ejecting from the generator and manually editing generated source files.

**Reply by sgentile** [reply]
What about using what you described here ?

https://prisma.typegraphql.com/docs/advanced/additional-decorators

**Reply by MichalLytek** [reply]
This won't let you add dataloaders

**Reply by sgentile** [reply]
But I'm still a bit confused, if you could help me understand better:

The models are generated, but I have to add directive attributes.   ie. 

```
@TypeGraphQL.Directive(`@key(fields: "id")`)
```

Is there not a way to add decorators to the generated models before the federated schema is generated ?

**Reply by sgentile** [reply]
Back to dataloader - in your example :

https://github.com/MichalLytek/type-graphql/blob/master/examples/apollo-federation/products/product-reference.ts

Would I just create a file like this and implement the dataloader there ?

ie. I do something similar with my NestJS code

```
@ResolveReference()
  async resolveReference(reference: { __typename: string; id: string }) {
    const user = await this.userDataLoader.batchUsers.load(reference.id);
    return user;
  }
```

that is part of the resolver - so I'd need to be able to extend the resolver ?

**Reply by MichalLytek** [reply]
> Is there not a way to add decorators to the generated models before the federated schema is generated ?

1) You run the generator which creates classes
2) You import those classes in you code
3) You call the apply decorators function to add directives to types
4) You call `buildSchema` AFTER so that it can see added decorators

**Reply by MichalLytek** [reply]
> that is part of the resolver - so I'd need to be able to extend the resolver ?

Resolve reference is added manually to the schema:
https://github.com/MichalLytek/type-graphql/blob/master/examples/apollo-federation/products/index.ts#L8-L17

**Reply by sgentile** [reply]
So I need to add the following to my 'Tools' model:

@TypeGraphQL.Directive(`@key(fields: "id")`)

```
@TypeGraphQL.ObjectType({
  isAbstract: true,
})
@TypeGraphQL.Directive(`@key(fields: "id")`)
export class Tool {
```

I don't see which 'apply decorators' I'd use here - is shows 'resolvers', 'fields' but this is a model type ?
https://prisma.typegraphql.com/docs/advanced/additional-decorators

ie. applyResolversEnhanceMap and ResolversEnhanceMap - but i need to enhance the generated Model itself not the resolver ?

**Reply by MichalLytek** [reply]
Can you scroll down a bit before asking?

![image](https://user-images.githubusercontent.com/10618781/136583696-db7633b7-6566-4a7a-848b-ef6ef9b20e65.png)

**Reply by sgentile** [reply]
Those are decorating the fields, I need to decorate the class ?   The federation 'key' is applied to the class not a field ???

```
@TypeGraphQL.Directive(`@key(fields: "id")`)
export class Tool {
```

**Reply by MichalLytek** [reply]
I will enlarge a little bit more:
![image](https://user-images.githubusercontent.com/10618781/136584812-3bcfb7b4-1ade-4a6d-9965-36be7e1f879d.png)

**Reply by sgentile** [reply]
ok sorry Michael, I think I now see the 'class' options - doh 

```
const postEnhanceConfig: ModelConfig<"Tool"> = {
  class: [TypeGraphQL.Directive(`@key(fields: "id")`)],
};
const modelsEnhanceMap: ModelsEnhanceMap = {
  Tool: postEnhanceConfig,
  // or apply it inline
  // Recipe: {
  //   class: [
  //     // decorators for @ObjectType model class
  //     Extensions({ logMessage: "Secret recipe", logLevel: LogLevel.INFO }),
  //   ],
  //   fields: {
  //     // decorator for model class fields
  //     averageRating: [Authorized()],
  //   },
  // },
};
```

**Reply by sgentile** [reply]
I added that but the generated Tool model is not showing that attribute

```
@TypeGraphQL.ObjectType({
  isAbstract: true
})
export class Tool {
```

**Reply by MichalLytek** [reply]
It's added in memory to the classes, not to the generated code.

**Reply by sgentile** [reply]
```
export type ModelConfig<TModel extends ModelNames> = {
  class?: ClassDecorator[];
  fields?: ModelFieldsConfig<TModel>;
};

export type ModelsEnhanceMap = {
  [TModel in ModelNames]?: ModelConfig<TModel>;
};
```
![image](https://user-images.githubusercontent.com/132404/136586411-5f716e3b-66f7-4746-a7b4-e00397552c64.png)

**Reply by sgentile** [reply]
> It's added in memory to the classes, not to the generated code.

any way I can confirm it's added ?

**Reply by MichalLytek** [reply]
Oh my god, dude! Do you know that `Recipe` is the name of the object type class from example? 😅

**Reply by MichalLytek** [reply]
I'm really sorry but I don't have time for leading you by hand. You need to ask community for the help.

**Reply by sgentile** [reply]
Last question I hope for your sake....

So using the federation example:
https://github.com/MichalLytek/type-graphql/blob/master/examples/apollo-federation/products/product-reference.ts

Any recommendations on getting access to the prisma instance from within the reference ?

```
import { Tool } from "../prisma/generated/type-graphql/models";

export async function resolveToolReference(
  reference: Pick<Tool, "id">
): Promise<Tool | undefined> {
 
  return undefined;
}
```

I've added the reference to my index:
```
const postEnhanceConfig: ModelConfig<"Tool"> = {
  class: [TypeGraphQL.Directive(`@key(fields: "id")`)],
};
const modelsEnhanceMap: ModelsEnhanceMap = {
  Tool: postEnhanceConfig,
};

async function main() {
  applyModelsEnhanceMap(modelsEnhanceMap);
  const schema = await buildFederatedSchema(
    {
      resolvers: [...resolvers],
      orphanedTypes: [Tool],
    },
    {
      Tool: { __resolveReference: resolveToolReference },
    }
  );
  const prisma = new PrismaClient();
  const server = new ApolloServer({
    schema,
    context: (): Context => ({ prisma }),
  });
  const { port } = await server.listen(4002);
  console.log(`GraphQL is listening on ${port}!`);
}
```

**Reply by sgentile** [reply]
> Oh my god, dude! Do you know that `Recipe` is the name of the object type class from example? 😅

ugh sorry Michael

**Comment by sgentile** [comment]
So I was able to make partial progress.   Hopefully the *community* can provide some help here ... 

There are 2 additional requirements to make Federation work.   Adding a new type, extending it and applying the @external attribute. (User is not part of the Prisma model)

```
type Post @key(fields: "id") {
  id: ID!
  title: String!
  content: String!
  published: Boolean!
  createdAt: String!
  userId: Int
  user: User
}

extend type User @key(fields: "id") {
  id: ID! @external
  posts: [Post]
}
```

Attempted code.   

```
@TypeGraphQL.ObjectType({
  isAbstract: true,
})
export class User {
  @TypeGraphQL.Field((_type) => TypeGraphQL.Int, {
    nullable: false,
  })
  id!: number;
}

const userEnhanceConfig: ModelConfig<"User"> = {
  class: [TypeGraphQL.Directive(`@key(fields: "id")`)],
};

const postEnhanceConfig: ModelConfig<"Post"> = {
  class: [TypeGraphQL.Directive(`@key(fields: "id")`)],
};
const modelsEnhanceMap: ModelsEnhanceMap = {
  Post: postEnhanceConfig,
  User: userEnhanceConfig,
};

async function main() {
  applyModelsEnhanceMap(modelsEnhanceMap);
  const schema = await buildFederatedSchema(
    {
      resolvers: [...resolvers],
      orphanedTypes: [User],
    },
    {
      Post: { __resolveReference: resolveToolReference },
    }
  );
  const server = new ApolloServer({
    schema,
    context: (): Context => ({ prisma: prisma() }),
  });
  const { port } = await server.listen(4000);
  console.log(`GraphQL is listening on ${port}!`);
```

Errors:

Type '"User"' does not satisfy the constraint '"Post"'.ts(2344)
 -- and --
Type '{ Post: ModelConfig<"Post">; User: ModelConfig<"User">; }' is not assignable to type 'ModelsEnhanceMap'.
  Object literal may only specify known properties, and 'User' does not exist in type 'ModelsEnhanceMap'.ts(2322)
